src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter.cs

```diff
@@ -24,12 +37,45 @@ public void OnTempDataSaving(ITempDataDictionary tempData)
                     var originalValue = kvp.Value;
 
                     var newValue = property.GetValue(Subject);
-                    if (newValue != null && newValue != originalValue)
+                    if (newValue != null && !newValue.Equals(originalValue))
                     {
                         tempData[Prefix + property.Name] = newValue;
                     }
                 }
             }
         }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            Subject = context.Controller;
+            var tempData = _factory.GetTempData(context.HttpContext);
+            var properties = PropertyHelpers.Select(p => p.Property).ToArray();
```

Comment:
Let's avoid the Linq call entirely. Also, this needs to throw if `PropertyHelpers` is not initialized.